### ISSUE TITLE
[ROCm][CK][gfx1201][Windows] FlashAttention backward is 4.4–7.4x slower than PyTorch SDPA on RX 9070 XT

### ISSUE THREAD

**Issue by qweqweewqe7-create** [Dao-AILab/flash-attention#2800]
## Summary

I am seeing unexpectedly poor backward performance from the ROCm CK backend of FlashAttention on an AMD Radeon RX 9070 XT (gfx1201, RDNA4) on Windows.

Forward performance is good and consistently faster than PyTorch SDPA, but backward is **4.4x to 7.4x slower** than SDPA, with the gap increasing at longer sequence lengths. Forward + backward is therefore **3.3x to 6.0x slower** than SDPA.

Numerical results look correct, so this appears to be a performance issue rather than a correctness issue.

This resembles the old ROCm FlashAttention backward performance issue:
https://github.com/ROCm/flash-attention/issues/44

That issue was eventually reported as fixed after the ck_tile refactor. I am wondering whether the gfx1201 / RDNA4 backward path is hitting an untuned or fallback CK kernel.

## Environment

- OS: Windows 11 Pro
- GPU: AMD Radeon RX 9070 XT
- Architecture: gfx1201
- PyTorch: `2.15.0a0+rocm10.1.0a20260814`
- HIP: `7.16.26323`
- FlashAttention source tag: `fa4-v4.0.0.beta26`
- FlashAttention Python package version: `2.8.4`
- Composable Kernel submodule: `c56c6750d0fc54ed771d532cc92c316423449614`
- `rocm-sdk-core`: `10.1.0a20260814`
- `rocm-sdk-devel`: `10.1.0a20260814`
- `rocm-sdk-device-gfx1201`: `10.1.0a20260814`
- AMD clang: `23.0.0git`, commit `a01cdbd92d1f931738a4897502747a68ae36eccf`
- MSVC: 14.44 / cl 19.44

Build configuration:

```text
BUILD_TARGET=rocm
FLASH_ATTENTION_FORCE_BUILD=TRUE
GPU_ARCHS=gfx1201
PYTORCH_ROCM_ARCH=gfx1201
OPT_DIM=32,64,128,256
MAX_JOBS=8
```

CK backend was used, not the Triton AMD backend.

On this dual-GPU system I exposed the RX 9070 XT as the only visible GPU for the benchmark, so the benchmark device is logical `cuda:0`.

## Benchmark configuration

- dtype: `torch.bfloat16`
- batch: 1
- heads: 16
- head_dim: 128
- causal: False

### Forward only

| Sequence | PyTorch SDPA | Flash CK | Flash / SDPA |
|---:|---:|---:|---:|
| 1024 | 0.267 ms | 0.134 ms | 0.503x |
| 2048 | 0.610 ms | 0.439 ms | 0.719x |
| 4096 | 2.011 ms | 1.450 ms | 0.721x |

Flash CK forward is faster than SDPA as expected.

### Backward only

| Sequence | PyTorch SDPA | Flash CK | Flash / SDPA |
|---:|---:|---:|---:|
| 1024 | 1.068 ms | 4.690 ms | 4.391x |
| 2048 | 2.738 ms | 17.660 ms | 6.449x |
| 4096 | 9.357 ms | 69.484 ms | 7.426x |

The backward gap becomes significantly larger with sequence length.

### Forward + backward

| Sequence | PyTorch SDPA | Flash CK | Flash / SDPA |
|---:|---:|---:|---:|
| 1024 | 1.487 ms | 4.940 ms | 3.321x |
| 2048 | 3.490 ms | 18.268 ms | 5.234x |
| 4096 | 11.950 ms | 71.421 ms | 5.976x |

Peak allocated memory was effectively identical in this microbenchmark.

## Correctness

For `B=1, S=1024, H=16, D=128`, BF16:

```text
Flash CK vs PyTorch SDPA forward relative L2: 2.826219e-03

dQ relative L2: 1.489228e-03
dK relative L2: 1.518024e-03
dV relative L2: 1.391158e-04
```

So the outputs and gradients appear numerically reasonable.

## Real workload observation

I see the same qualitative behavior in an actual LoRA training workload: the CK FlashAttention path is substantially slower than PyTorch SDPA during training, despite being slightly faster for inference/sampling.

## Minimal reproduction

On a multi-GPU system, expose only the target GPU before running. For example, in PowerShell on my system:

```powershell
$env:CUDA_VISIBLE_DEVICES = "1"
python repro_rocm_ck_bwd_gfx1201.py
```

Reproduction script:

```python
import gc
import statistics
import time

import torch
import torch.nn.functional as F
from flash_attn import flash_attn_func

DEVICE = torch.device("cuda:0")
DTYPE = torch.bfloat16
BATCH = 1
HEADS = 16
HEAD_DIM = 128
SEQS = (1024, 2048, 4096)

WARMUP_FWD = 10
ITERS_FWD = 50
ROUNDS_FWD = 7
WARMUP_BWD = 3
ITERS_BWD = 10
ROUNDS_BWD = 5
WARMUP_TRAIN = 3
ITERS_TRAIN = 10
ROUNDS_TRAIN = 5


def sync():
    torch.cuda.synchronize(DEVICE)


def cleanup():
    gc.collect()
    torch.cuda.empty_cache()
    sync()


def sdpa(q, k, v):
    qh = q.transpose(1, 2)
    kh = k.transpose(1, 2)
    vh = v.transpose(1, 2)
    out = F.scaled_dot_product_attention(
        qh, kh, vh, dropout_p=0.0, is_causal=False
    )
    return out.transpose(1, 2)


def flash_ck(q, k, v):
    return flash_attn_func(
        q, k, v,
        dropout_p=0.0,
        softmax_scale=None,
        causal=False,
    )


def rel_l2(a, b):
    a = a.float()
    b = b.float()
    return ((a - b).norm() / b.norm().clamp_min(1e-12)).item()


def median_rounds(run_one_round, rounds):
    vals = [run_one_round() for _ in range(rounds)]
    return statistics.median(vals), statistics.mean(vals), min(vals)


def bench_forward(fn, B, S, H, D):
    cleanup()
    q = torch.randn(B, S, H, D, device=DEVICE, dtype=DTYPE)
    k = torch.randn(B, S, H, D, device=DEVICE, dtype=DTYPE)
    v = torch.randn(B, S, H, D, device=DEVICE, dtype=DTYPE)

    with torch.inference_mode():
        for _ in range(WARMUP_FWD):
            fn(q, k, v)
        sync()

        def one_round():
            sync()
            t0 = time.perf_counter()
            for _ in range(ITERS_FWD):
                fn(q, k, v)
            sync()
            return (time.perf_counter() - t0) * 1000.0 / ITERS_FWD

        return median_rounds(one_round, ROUNDS_FWD)


def bench_backward_only(fn, B, S, H, D):
    cleanup()
    q = torch.randn(B, S, H, D, device=DEVICE, dtype=DTYPE, requires_grad=True)
    k = torch.randn(B, S, H, D, device=DEVICE, dtype=DTYPE, requires_grad=True)
    v = torch.randn(B, S, H, D, device=DEVICE, dtype=DTYPE, requires_grad=True)

    out = fn(q, k, v)
    grad = torch.randn_like(out)

    for _ in range(WARMUP_BWD):
        q.grad = None
        k.grad = None
        v.grad = None
        out.backward(grad, retain_graph=True)
    sync()

    def one_round():
        sync()
        t0 = time.perf_counter()
        for _ in range(ITERS_BWD):
            q.grad = None
            k.grad = None
            v.grad = None
            out.backward(grad, retain_graph=True)
        sync()
        return (time.perf_counter() - t0) * 1000.0 / ITERS_BWD

    return median_rounds(one_round, ROUNDS_BWD)


def bench_forward_backward(fn, B, S, H, D):
    cleanup()
    q = torch.randn(B, S, H, D, device=DEVICE, dtype=DTYPE, requires_grad=True)
    k = torch.randn(B, S, H, D, device=DEVICE, dtype=DTYPE, requires_grad=True)
    v = torch.randn(B, S, H, D, device=DEVICE, dtype=DTYPE, requires_grad=True)

    def step():
        q.grad = None
        k.grad = None
        v.grad = None
        out = fn(q, k, v)
        out.float().square().mean().backward()

    for _ in range(WARMUP_TRAIN):
        step()
    sync()

    def one_round():
        sync()
        t0 = time.perf_counter()
        for _ in range(ITERS_TRAIN):
            step()
        sync()
        return (time.perf_counter() - t0) * 1000.0 / ITERS_TRAIN

    return median_rounds(one_round, ROUNDS_TRAIN)


def accuracy_check():
    B, S, H, D = 1, 1024, HEADS, HEAD_DIM
    torch.manual_seed(1234)

    q = torch.randn(B, S, H, D, device=DEVICE, dtype=DTYPE)
    k = torch.randn(B, S, H, D, device=DEVICE, dtype=DTYPE)
    v = torch.randn(B, S, H, D, device=DEVICE, dtype=DTYPE)

    with torch.no_grad():
        ref = sdpa(q, k, v)
        out = flash_ck(q, k, v)

    print("[correctness]")
    print(f"  forward rel L2: {rel_l2(out, ref):.6e}")

    q1 = q.detach().clone().requires_grad_(True)
    k1 = k.detach().clone().requires_grad_(True)
    v1 = v.detach().clone().requires_grad_(True)
    q2 = q.detach().clone().requires_grad_(True)
    k2 = k.detach().clone().requires_grad_(True)
    v2 = v.detach().clone().requires_grad_(True)

    ref = sdpa(q1, k1, v1)
    out = flash_ck(q2, k2, v2)
    grad = torch.randn_like(ref)
    ref.backward(grad)
    out.backward(grad)
    sync()

    print(f"  dQ rel L2     : {rel_l2(q2.grad, q1.grad):.6e}")
    print(f"  dK rel L2     : {rel_l2(k2.grad, k1.grad):.6e}")
    print(f"  dV rel L2     : {rel_l2(v2.grad, v1.grad):.6e}")
    print()


def main():
    if not torch.cuda.is_available():
        raise RuntimeError("HIP/CUDA device is not available")

    torch.cuda.set_device(0)
    props = torch.cuda.get_device_properties(0)

    print("=" * 80)
    print("ROCm CK FlashAttention backward performance repro")
    print("=" * 80)
    print("torch :", torch.__version__)
    print("HIP   :", torch.version.hip)
    print("GPU   :", torch.cuda.get_device_name(0))
    print("arch  :", getattr(props, "gcnArchName", "unknown"))
    print("dtype :", DTYPE)
    print()

    accuracy_check()

    for S in SEQS:
        print("=" * 80)
        print(f"B={BATCH}, S={S}, H={HEADS}, D={HEAD_DIM}")
        print("=" * 80)

        s_fwd = bench_forward(sdpa, BATCH, S, HEADS, HEAD_DIM)[0]
        f_fwd = bench_forward(flash_ck, BATCH, S, HEADS, HEAD_DIM)[0]
        s_bwd = bench_backward_only(sdpa, BATCH, S, HEADS, HEAD_DIM)[0]
        f_bwd = bench_backward_only(flash_ck, BATCH, S, HEADS, HEAD_DIM)[0]
        s_train = bench_forward_backward(sdpa, BATCH, S, HEADS, HEAD_DIM)[0]
        f_train = bench_forward_backward(flash_ck, BATCH, S, HEADS, HEAD_DIM)[0]

        print(f"forward       SDPA={s_fwd:9.3f} ms  FlashCK={f_fwd:9.3f} ms  ratio={f_fwd/s_fwd:6.3f}x")
        print(f"backward      SDPA={s_bwd:9.3f} ms  FlashCK={f_bwd:9.3f} ms  ratio={f_bwd/s_bwd:6.3f}x")
        print(f"forward+bwd   SDPA={s_train:9.3f} ms  FlashCK={f_train:9.3f} ms  ratio={f_train/s_train:6.3f}x")
        print()


if __name__ == "__main__":
    main()
```

## Question

Is this level of backward slowdown expected on gfx1201 / RDNA4?

Could the gfx1201 backward path be selecting a generic/fallback CK kernel, or is backward currently not tuned for RDNA4?

I can provide additional benchmarks, generated kernel information, or profiling data if useful.
